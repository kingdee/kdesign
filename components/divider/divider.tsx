import React, { FunctionComponentElement, useContext, useEffect, useMemo } from 'react'
import classNames from 'classnames'
import ConfigContext from '../config-provider/ConfigContext'
import { getCompProps } from '../_utils'
import devWarning from '../_utils/devwarning'
import { tuple } from '../_utils/type'

export const DividerTypes = tuple('horizontal', 'vertical')
export type DividerType = typeof DividerTypes[number]

export const DividerOrientations = tuple('left', 'center', 'right')
export type DividerOrientation = typeof DividerOrientations[number]

export const DividerBorderStyles = tuple('solid', 'dashed', 'dotted')
export type DividerBorderStyle = typeof DividerBorderStyles[number]

export interface IDividerProps {
  style?: Record<string, unknown> // 内联样式
  className?: string // 样式名
  prefixCls?: string // 样式前缀
  children?: React.ReactNode // 子元素
  borderStyle?: DividerBorderStyle // 线段样式
  type?: DividerType // 线段类型
  orientation?: DividerOrientation // 内嵌子元素布局
  orientationMargin?: number | string // 内嵌子元素具体位置
}

const Divider = (props: IDividerProps): FunctionComponentElement<IDividerProps> => {
  const { getPrefixCls, prefixCls, compDefaultProps: userDefaultProps } = useContext(ConfigContext)
  const dividerProps = getCompProps('Divider', userDefaultProps, props)
  const {
    className,
    prefixCls: customPrefixcls,
    children,
    type,
    orientationMargin,
    orientation,
    borderStyle,
    ...others
  } = dividerProps
  const dividerPrefixCls = getPrefixCls!(prefixCls, 'divider', customPrefixcls)
  const orientationPrefix = orientation.length > 0 ? `-${orientation}` : orientation
  const hasChildren = !!children
  const isHorizontal = type === 'horizontal'
  const isVertical = type === 'vertical'
  const optimizedOrientationMargin = useMemo<string | number>(() => {
    if (/^[0-9]+$/.test(orientationMargin)) {
      return orientationMargin + 'px'
    }
    return ''
  }, [orientationMargin])
  const dividerClass = classNames(dividerPrefixCls, className, {
    [`${dividerPrefixCls}-with-text`]: hasChildren && isHorizontal,
    [`${dividerPrefixCls}-with-text${orientationPrefix}`]: hasChildren && isHorizontal,
    [`${dividerPrefixCls}-solid`]: borderStyle === 'solid',
    [`${dividerPrefixCls}-dashed`]: borderStyle === 'dashed',
    [`${dividerPrefixCls}-dotted`]: borderStyle === 'dotted',
    [`${dividerPrefixCls}-vertical`]: isVertical,
    [`${dividerPrefixCls}-margin-left`]: isHorizontal && orientation === 'left' && optimizedOrientationMargin,
    [`${dividerPrefixCls}-margin-right`]: isHorizontal && orientation === 'right' && optimizedOrientationMargin,
  })
  const innerTextClass = classNames(`${dividerPrefixCls}-inner-text`)

  const innerTextStyle: React.CSSProperties = {
    ...(orientation === 'left' && { marginLeft: optimizedOrientationMargin }),
    ...(orientation === 'right' && { marginRight: optimizedOrientationMargin }),
  }

  useEffect(() => {
    devWarning(isVertical && children, 'divider', '`children` not working in `vertical` mode.')
  }, [children, isVertical])

  return (
    <div className={dividerClass} {...others}>
      {isHorizontal && hasChildren && (
        <div className={innerTextClass} style={innerTextStyle}>
          {children}
        </div>
      )}
    </div>
  )
}

Divider.displayName = 'Divider'
export default Divider
