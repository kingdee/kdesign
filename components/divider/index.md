---
category: Components
type: 布局
title: Divider
subtitle: 分割线
---

分割线是一种用于在界面中创建水平或垂直分隔元素的线条

## 使用场景

• 当用户需要对内容进行划分时

## API

| 属性 | 说明 | 类型 | 默认值 | 可选值 | 版本 |
| --- | --- | --- | --- | --- | --- |
| borderStyle | 分割线样式 | string | `solid` | `solid` \| `dashed` \| `dotted` | 2.0.0 |
| children | 嵌套的标题内容 | ReactNode | - | - | 2.0.0 |
| className | 分割线样式类名 | string | - | - | 2.0.0 |
| orientation | 分割线标题的位置 | string | `left` | `left` \| `center` \| `right` | 2.0.0 |
| orientationMargin | 标题和最近 left/right 边框之间的距离，去除了分割线，同时 orientation 必须为 left 或 right。如果传入 string 类型的数字且不带单位，默认单位是 px | string \| number | - | - | 2.0.0 |
| style | 分割线样式对象 | CSSProperties | - | - | 2.0.0 |
| type | 水平还是垂直类型 | string | `horizontal` | `horizontal` \| `vertical` | 2.0.0 |


## Design Token

| 分类 | 组件token | 全局token | 默认值 |
| --- | --- | --- | --- |
| color | --kd-c-divider-color-text | --kd-g-color-text-primary | #212121 |
| font | --kd-c-divider-font-size | --kd-g-font-size-small | 12px |
| spacing | --kd-c-divider-spacing-padding-horizontal | - | 0px |
