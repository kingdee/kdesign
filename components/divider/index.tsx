import Divider from './divider'
export * from './divider'
export default Divider
