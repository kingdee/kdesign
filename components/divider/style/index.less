@import '../../style/themes/index';
@import './mixin.less';
@divider-prefix-cls: ~'@{kd-prefix}-divider';

.@{divider-prefix-cls} {
  .divider;

  width: 900px;
  padding: @divider-spacing-padding-horizontal;
  border-block-start: 1px solid @color-border-strong;

  &-vertical {
    display: inline-block;
    border-inline-start: 1px solid @color-border-strong;
    width: 0;
    height: 1em;
  }

  &-solid {
    .borderStyle(solid)
  }

  &-dashed {
    .borderStyle(dashed)
  }

  &-dotted {
    .borderStyle(dotted)
  }

  &-with-text {
    border: unset;

    .@{divider-prefix-cls}-inner-text {
      display: flex;
      align-items: center;

      &::before {
        content: '';
        width: 74px;
        margin-right: 8px;
        border-block-start: 1px solid @color-border-strong;
      }

      &::after {
        content: '';
        width: calc(100% - 74px);
        margin-left: 8px;
        border-block-start: 1px solid @color-border-strong;
      }
    }
  }

  &-with-text-center {
    .@{divider-prefix-cls}-inner-text {

      &::before,
      &::after {
        width: 50%;
      }
    }
  }

  &-with-text-right {
    .@{divider-prefix-cls}-inner-text {

      &::before {
        width: calc(100% - 74px);

      }

      &::after {
        width: 74px;
      }
    }
  }

  &-margin-left {
    .@{divider-prefix-cls}-inner-text {

      &::before {
        width: 0;

      }

      &::after {
        width: 100%;
      }
    }
  }

  &-margin-right {
    .@{divider-prefix-cls}-inner-text {

      &::before {
        width: 100%;

      }

      &::after {
        width: 0;
      }
    }
  }
}
