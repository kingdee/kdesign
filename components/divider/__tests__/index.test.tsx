import { mount, render } from 'enzyme'
import React from 'react'
import mountTest from '../../../tests/shared/mountTest'
import ConfigProvider from '../../config-provider/index'
import { DividerTypes, DividerOrientations, DividerBorderStyles } from '../divider'
import Divider from '../index'

describe('Divider', () => {
  // 1. mount test
  describe('1. mount test', () => {
    mountTest(Divider)
    DividerTypes.forEach((type) => {
      mountTest(() => <Divider type={type} />)
    })
    DividerOrientations.forEach((orientation) => {
      mountTest(() => <Divider orientation={orientation} />)
    })
    DividerBorderStyles.forEach((borderStyle) => {
      mountTest(() => <Divider borderStyle={borderStyle} />)
    })
  })

  // 2. render test
  describe('2. render test', () => {
    it('renders correctly', () => {
      expect(render(<Divider />)).toMatchSnapshot()

      expect(render(<Divider style={{ width: 100 }} />)).toMatchSnapshot()

      expect(render(<Divider orientationMargin="12">Divider</Divider>)).toMatchSnapshot()

      DividerTypes.forEach((type) => {
        expect(render(<Divider type={type} />)).toMatchSnapshot()
      })

      DividerOrientations.forEach((orientation) => {
        expect(render(<Divider orientation={orientation} />)).toMatchSnapshot()
      })

      DividerBorderStyles.forEach((borderStyle) => {
        expect(render(<Divider borderStyle={borderStyle} />)).toMatchSnapshot()
      })
    })
  })

  // 3. warns in component
  describe('3. warns in component', () => {
    it('warns if type is vertical and children exist', () => {
      const mockWarn = jest.fn()
      jest.spyOn(console, 'warn').mockImplementation(mockWarn)
      mount(<Divider type="vertical">divider</Divider>)
      expect(mockWarn).toHaveBeenCalledTimes(1)
      expect(mockWarn.mock.calls[0][0]).toMatch(
        'Warning: [kdesign]-divider: `children` not working in `vertical` mode.',
      )
    })
  })

  // 4. render null or undefined without errors
  describe('4. render null or undefined without errors', () => {
    it('render null or undefined without errors', () => {
      const wrapper = (
        <Divider>
          {null}
          {undefined}
        </Divider>
      )
      expect(wrapper).toMatchSnapshot()
    })
  })

  // 5. displayName
  describe('5. displayName', () => {
    it('should have displayName static property', () => {
      expect(Divider.displayName).toBe('Divider')
    })
  })

  // 6. class state
  describe('6. class state', () => {
    it('other properties should be passed', () => {
      const wrapper = mount(
        <Divider className="my-class" data-test="test">
          hello world
        </Divider>,
      )
      expect(wrapper).toHaveClassName('my-class')
      expect(wrapper.prop('data-test')).toEqual('test')

      const defaultDivider = mount(<Divider />)
      expect(defaultDivider.find('.kd-divider')).toHaveClassName('kd-divider-solid')

      const verticalDivider = mount(<Divider type="vertical" />)
      expect(verticalDivider.find('.kd-divider')).toHaveClassName('kd-divider-vertical')

      DividerBorderStyles.forEach((borderStyle) => {
        const testBorderStyleDivider = mount(<Divider borderStyle={borderStyle}></Divider>)
        expect(testBorderStyleDivider.find(`.kd-divider`)).toHaveClassName(`kd-divider-${borderStyle}`)
      })

      DividerOrientations.forEach((orientation) => {
        const testOrientationDivider = mount(<Divider orientation={orientation}>Divider</Divider>)
        expect(testOrientationDivider.find(`.kd-divider`)).toHaveClassName(`kd-divider-with-text`)
        expect(testOrientationDivider.find(`.kd-divider`)).toHaveClassName(`kd-divider-with-text-${orientation}`)
      })
    })
  })

  // 7.component interaction(event)

  // 8.config provider
  describe('8.config provider', () => {
    it('should config use config provider', () => {
      const clipboardConfig = {
        compDefaultProps: {
          Divider: {
            borderStyle: 'dotted',
            orientation: 'right',
          },
        },
      }
      const wrapper = mount(
        <ConfigProvider value={clipboardConfig}>
          <Divider>Divider</Divider>
        </ConfigProvider>,
      )
      expect(wrapper.find('.kd-divider')).toHaveClassName('kd-divider-dotted')
      expect(wrapper.find(`.kd-divider`)).toHaveClassName('kd-divider-with-text')
      expect(wrapper.find(`.kd-divider`)).toHaveClassName('kd-divider-with-text-right')
    })
  })

  // 9. ref test

  // 10. api test
  describe('10. api test', () => {
    it('orientationMargin api', () => {
      const wrapper = mount(
        <Divider orientation="left" orientationMargin="20">
          divider
        </Divider>,
      )
      expect(wrapper.find('.kd-divider').find('.kd-divider-inner-text')).toExist()
      expect(wrapper.find('.kd-divider-inner-text')).toHaveStyle({ marginLeft: '20px' })
    })
  })

  // 11. special case
})
