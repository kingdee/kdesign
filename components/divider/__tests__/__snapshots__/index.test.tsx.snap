// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider 2. render test renders correctly 1`] = `
<div
  class="kd-divider kd-divider-solid"
/>
`;

exports[`Divider 2. render test renders correctly 2`] = `
<div
  class="kd-divider kd-divider-solid"
  style="width:100px"
/>
`;

exports[`Divider 2. render test renders correctly 3`] = `
<div
  class="kd-divider kd-divider-with-text kd-divider-with-text-left kd-divider-solid kd-divider-margin-left"
>
  <div
    class="kd-divider-inner-text"
    style="margin-left:12px"
  >
    Divider
  </div>
</div>
`;

exports[`Divider 2. render test renders correctly 4`] = `
<div
  class="kd-divider kd-divider-solid"
/>
`;

exports[`Divider 2. render test renders correctly 5`] = `
<div
  class="kd-divider kd-divider-solid kd-divider-vertical"
/>
`;

exports[`Divider 2. render test renders correctly 6`] = `
<div
  class="kd-divider kd-divider-solid"
/>
`;

exports[`Divider 2. render test renders correctly 7`] = `
<div
  class="kd-divider kd-divider-solid"
/>
`;

exports[`Divider 2. render test renders correctly 8`] = `
<div
  class="kd-divider kd-divider-solid"
/>
`;

exports[`Divider 2. render test renders correctly 9`] = `
<div
  class="kd-divider kd-divider-solid"
/>
`;

exports[`Divider 2. render test renders correctly 10`] = `
<div
  class="kd-divider kd-divider-dashed"
/>
`;

exports[`Divider 2. render test renders correctly 11`] = `
<div
  class="kd-divider kd-divider-dotted"
/>
`;

exports[`Divider 4. render null or undefined without errors render null or undefined without errors 1`] = `<Divider />`;
